store/snap_remote_repo.go

```diff
@@ -232,6 +257,111 @@ func (s *SnapUbuntuStoreRepository) checkStoreResponse(resp *http.Response) {
 	}
 }
 
+func getRequiresPurchase(prices map[string]float64, purchases purchaseList) bool {
+	// if the snap is free, then it doesn't need purchasing
+	if len(prices) == 0 {
+		return false
+	}
+
+	for _, purchase := range purchases {
+		// if the purchase is not an in-app purchase
+		if purchase.ItemSKU == "" {
+			return false
+		}
+	}
+
+	// the snap is not free, and we couldn't find a purchase
+	return true
+}
+
+func (s *SnapUbuntuStoreRepository) getPurchases(name string, auther Authenticator) (purchaseList, error) {
+	var purchases purchaseList
+
+	// only build the request if we have an Authenticator
+	if auther != nil {
+		purchasesURL, err := s.purchasesURI.Parse(name + "/")
+		if err != nil {
+			return nil, err
+		}
+
+		q := purchasesURL.Query()
+		q.Set("include_item_purchases", "true")
+		purchasesURL.RawQuery = q.Encode()
+
+		req, err := http.NewRequest("GET", purchasesURL.String(), nil)
+		if err != nil {
+			return nil, err
+		}
+
+		s.applyUbuntuStoreHeaders(req, "application/json", auther)
+
+		resp, err := s.client.Do(req)
+		if err != nil {
+			return nil, err
+		}
+		defer resp.Body.Close()
+
+		switch {
+		case resp.StatusCode == http.StatusNotFound:
+			break
+		case resp.StatusCode == http.StatusUnauthorized:
+			return nil, decodeUnauthorizedError(resp.Body)
+		case resp.StatusCode != http.StatusOK:
+			return nil, fmt.Errorf("SnapUbuntuStoreRepository: unexpected HTTP status code %d while looking for snap purcahses: %q", resp.StatusCode, name)
+		default:
+			dec := json.NewDecoder(resp.Body)
+			if err := dec.Decode(&purchases); err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	return purchases, nil
+}
+
+func (s *SnapUbuntuStoreRepository) getAllPurchases(auther Authenticator) (map[string]purchaseList, error) {
```

Comment:
This is almost a copy of the method above. We can accept an empty name instead and tweak the one line accordingly, so it's all one method only.